scripts/backout_data_collection.py

```diff
@@ -0,0 +1,219 @@
+
+    with open(destination_filepath, 'w') as file:
+        file.write(json_data)
+
+    logger.info(f"Data successfully saved to {destination_filepath}")
+
+def preprocess_commits():
+    """
+    Preprocess all commit hashes to create a dictionary mapping the first
+    12 characters to the entire hash.
+
+    Args:
+        None
+    
+    Returns:
+        commit_dict (dict): dictionary of commit hashes
+    
+    """
+
+    commit_dict = {}
+    for commit in repository.get_commits(include_no_bug=False, include_backouts=True, include_ignored=False):
+        shortened_hash = commit['node'][:12]
+        commit_dict[shortened_hash] = commit['node']
+    return commit_dict
+
+
+def get_full_hash(shortened_hash: str, commit_dict) -> str:
+    """
+    Given a shortened hash of 12 characters, search in the commit_dict for the full hash and return it.
+
+    Args:
+        shortened_hash (str): 12 character long shortened hash of a commit
+        commit_dict (dict): Dictionary of shortened hashes to full hashes
+    Returns:
+        str: full 40 character long hash of commit, or None if not found
+    """
+
+    return commit_dict.get(shortened_hash)
+
+def find_full_hash(source_filepath: str, destination_filepath: str, commit_dict: dict) -> None:
+    """
+    Convert the first 12 characters of the hash in backsout (list) to the full 40 character long hash using commit_dict.
+
+    Args:
+        source_filepath (str): filepath to .json file containing raw backout data
+        destination_filepath (str): where to save the new .json file containing the full hashesh
+        commit_dict (dict): dictionary mapping shortened hashes to full hashes
+    Returns:
+        None
+    """
+
+    with open(source_filepath, 'r') as file:
+        data = json.load(file)
+
+    for commit in data:
+        for index, backout_commit in enumerate(commit['backsout']):
+            full_hash = get_full_hash(backout_commit, commit_dict)
+
+            # replace shortened hash with full hash
+            if full_hash:
+                commit['backsout'][index] = full_hash
+    
+    # save to a new file
+    with open(destination_filepath, 'w') as file:
+        json.dump(data, file, indent=4)
+        logger.info(f"Updated data has been saved to {destination_filepath}")
+    return data
+
+def clean_dataset(source_filepath: str, destination_filepath: str):
+    """
+    Given a source filepath to a .json file containing the full hashes, remove
+    unnecessary fields and save to destination_filepath.
+
+    Args:
+        source_filepath (str): filepath to .json file containing full hash data
+        destination_filepath (str): where to save new .json file containing cleaned data
+    Returns:
+        None
+    """
+
+    with open(source_filepath, 'r') as file:
+        data = json.load(file)
+    
+    filtered_data = []
+
+    # fields to keep, rest will be discarded
+    fields_to_keep = ['node', 'bug_id', 'desc', 'pushdate', 'backsout']
```

Comment:
The goal at the first stage is to have the following information for each case:
- The defect inducing commit
- The backout commit
- The fix commit